libsolidity/formal/SymbolicVariable.h

```diff
@@ -0,0 +1,71 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include <libsolidity/formal/SolverInterface.h>
+
+#include <libsolidity/ast/AST.h>
+
+#include <memory>
+
+namespace dev
+{
+namespace solidity
+{
+
+class Declaration;
+
+/**
+ * This class represents the symbolic version of a program variable.
+ */
+class SymbolicVariable
+{
+public:
+	explicit SymbolicVariable(Declaration const* _decl,
+		smt::SolverInterface& _interface);
+	SymbolicVariable(SymbolicVariable const&) = default;
+	SymbolicVariable(SymbolicVariable&&) = default;
+	SymbolicVariable& operator=(SymbolicVariable const&) = default;
+	SymbolicVariable& operator=(SymbolicVariable&&) = default;
+
+	smt::Expression operator()(int _seq) const
+	{
+		return valueAtSequence(_seq);
+	}
+
+	std::string uniqueSymbol() const;
+
+	/// Sets the var to the default value of its type.
+	virtual void setZeroValue(int _seq) = 0;
+	/// The unknown value depends on the type. For example, an interval is set for Integers.
```

Comment:
Is this actually true? For all types, the unknown value is just the restriction on any valid value, regardless of the type. For integers, this is no less an interval as for bools, it is the discrete set of all possible valid values.

Reply:
True. I wanted to give the example for the integers because this function won't really do anything for Bool, since the SMT solver already knows that. It doesn't know the same for integers, of course.

Reply:
Ah right, but I guess bool is the only exception there.